daiquiri/query/static/query/js/services/plot.js

```diff
@@ -1,208 +1,569 @@
-            }
-        } else {
-            service.errors.y = 'empty';
-        }
+    if (service.values.y3){
+      var y3_params = angular.extend({}, service.params, {
+        column: service.values.y3
+      });
+      res_promises.push(resources.rows.paginate(y3_params).$promise)
+    }
 
-        if (service.errors.x === null && service.errors.y === null) {
-            service.fetch().then(function() {
-                service.draw();
-                service.ready = true;
-            });
-        } else {
-            service.ready = true;
+    /* fill service.source with the results from the query*/
+    return $q.all(res_promises).then(function(results) {
+      service.source = new Bokeh.ColumnDataSource({
+        data: {
+          x: results[0].results,
+          y: results[1].results
         }
-    };
+      });
+      if (service.values.y2){service.source.data.y2 = results[2].results;}
+      if (service.values.y3){service.source.data.y3 = results[3].results;}
+    });
+  }
+
+  /* create a basic figure template */
+  create_figure = function(x_range, y_range){
+
+    var figure_options = {
+      width: 840,
+      height: 500,
+      x_range: x_range,
+      y_range: y_range,
+      plot_width: $('.col-md-9').width(),
+      tools: 'wheel_zoom,box_zoom,reset,save',
+      x_axis_label: service.labels.x,
+      y_axis_label: service.labels.y,
+      background_fill_color: '#f5f5f5'
+    }
+
+    if (service.webgl) {
+      figure_options['output_backend'] = 'webgl';
+    }
+
+    var figure = new Bokeh.Plotting.figure(figure_options);
+
+    figure.toolbar.active_scroll = figure.toolbar.wheel_zoom;
+    figure.outline_line_color = '#dddddd';
+    figure.toolbar.logo = null;
+    return figure;
+  }
+
+
+  get_xlimits = function(){
+    return [Math.min.apply(null, service.source.data.x),
+            Math.max.apply(null, service.source.data.x)]
+  }
 
-    service.clear = function() {
-        $('#canvas').empty();
+
+  get_ylimits = function(){
+    var ymin = Math.min.apply(null, service.source.data.y),
+        ymax = Math.max.apply(null, service.source.data.y);
+
+    if (service.source.data.y2){
+        ymin = Math.min.apply(null, service.source.data.y2.concat([ymin]));
+        ymax = Math.max.apply(null, service.source.data.y2.concat([ymax]));
     }
 
-    service.fetch = function() {
-        var x_params = angular.extend({}, service.params, {
-            column: service.values.x
+    if (service.source.data.y3){
+        ymin = Math.min.apply(null, service.source.data.y3.concat([ymin]));
+        ymax = Math.max.apply(null, service.source.data.y3.concat([ymax]));
+    }
+    return [ymin, ymax]
+  }
+
+
+  service.draw_scatter_plot = function() {
+    var xmin, xmax, ymin, ymax;
+    [xmin, xmax] = get_xlimits();
+    [ymin, ymax] = get_ylimits();
+
+    if (!isNaN(xmin) && !isNaN(xmax) && !isNaN(ymin) && !isNaN(ymax)) {
+
+      // compute a 1% padding around the data
```

Comment:
What about 10% ? 5% maybe? Did you tried?

Reply:
Changed to 5%. It looks much better than 1% used before.